File: problems/3-longest-substring-without-repeating-characters/sliding-window.cpp
/**
 * 使用集合：维护窗口中的字符集。
 */

#include <iostream>
#include <string>
#include <algorithm>
#include <unordered_set>
using namespace std;

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_set<char> exist;
        int j = 0;
        int res = 0;

        for (int i = 0; i < s.size(); ++i) {
            // 不合法：存在重复，则收敛(j跳转到上一个位置的右边)
            if (exist.count(s[i])) {
                for (; s[j] != s[i]; ++j) {
                    exist.erase(s[j]);
                }
                exist.erase(s[j++]);
            }
            
            // 合法
            exist.insert(s[i]);
            res = max(res, i - j + 1);
        }

        return res;
    }
};

File: 剑指offer/48/sliding-window.js
/**
 * 滑动窗口
 * 
 * 时间：O(N), 96ms
 * 空间：O(N), 38.2MB
 */
var lengthOfLongestSubstring = function (s) {
  const exist = new Set();
  let res = 0;
  let L = 0, R = 0;

  while (R < s.length) {
    const newChar = s[R++];

    // 保证窗口内不存在新字符
    while (exist.has(newChar)) {
      const oldChar = s[L++];
      exist.delete(oldChar);
    }

    exist.add(newChar);
    res = Math.max(res, R - L);
  }

  return res;
};